orttraining/tools/scripts/pipeline_model_split.py

```diff
@@ -0,0 +1,376 @@
+import sys
+import onnx
+from onnx import helper
+from onnx import TensorProto
+from onnx import OperatorSetIdProto
+
+# Edge that needs to be cut for the split.
+# If the edge is feeding into more than one nodes, and not all the nodes belong to the same cut,
+# specify those consuming nodes that need to be cut
+class CutEdge:
+    def __init__(self, edgeId, consumingNodes=None):
+        self.edgeId = edgeId
+        self.consumingNodes = consumingNodes
+
+def add_expand_shape(model, name, shape):
+    expand_edge = model.graph.value_info.add()
+    expand_edge.name = name
+    expand_edge.type.CopyFrom(shape)
+
+# Add wait/record/send/recv nodes and split the graph into disconnected subgraphs
+def split_graph(model, edgeIds):
+    ms_domain = 'com.microsoft'
+
+    # Add wait for initial inputs. This needs to be done first before new inputs
+    # are introduced from split
+    initializer_lists = [a.name for a in model.graph.initializer]
+    input_tensors = [
+        value.name for value in model.graph.input if value.name not in initializer_lists]
+
+    input_wait_signal = model.graph.input.add()
+    input_wait_signal.CopyFrom(helper.make_tensor_value_info(
+        'input_wait_signal', onnx.TensorProto.INT64, None))
+
+    input_wait = model.graph.node.add()
+    input_wait.CopyFrom(helper.make_node(
+        'WaitEvent',
+        inputs=['input_wait_signal'],
+        outputs=[],
+        domain=ms_domain))
+
+    for i in input_tensors:
+        for node in model.graph.node:
+            for j in range(len(node.input)):
+                if node.input[j] == i:
+                    print(node.input[j])
+                    node.input[j] = i + '_sync'
+
+    input_wait.input.extend(input_tensors)
+    input_wait.output.extend([i + '_sync' for i in input_tensors])
```

Comment:
Instead of editing every node using input_tensor, maybe better to change graph.input name to have _sync postfix?

Reply:
change graph.input name with _sync postfix won't be surficent, right? What we need to do is adding new edges from input_wait to the ops that were initially getting data directly from graph.input.

---
In reply to: [400687293](https://github.com/microsoft/onnxruntime/pull/3275#discussion_r400687293) [](ancestors = 400687293)

Reply:
I mean you may change the original graph input into a value_info, so all nodes won't need change. New edges will take new def name for input, and use name in original graph input as output.

---
In reply to: [401161112](https://github.com/microsoft/onnxruntime/pull/3275#discussion_r401161112) [](ancestors = 401161112,400687293)

Reply:
option 1: change original input = origional input name update + value_info update + new input/output edge update for input_waite node
option 2: keep original input = input consumer node's input update + new input/output edge update for input_waite node

I see both option works with roughly the same amount of work. If that's the case, I would leaning towards keeping the current implementation.

---
In reply to: [401237498](https://github.com/microsoft/onnxruntime/pull/3275#discussion_r401237498) [](ancestors = 401237498,401161112,400687293)

Reply:
Depends on the size of model, for really big ones like 96-layer BERT, editing graph input would be faster than nodes.

---
In reply to: [401272474](https://github.com/microsoft/onnxruntime/pull/3275#discussion_r401272474) [](ancestors = 401272474,401237498,401161112,400687293)

Reply:
We are only editing the nodes that are direct consumers of the inputs, so edited nodes will be the same number as the input if input and consumer node is 1-1 mapping. 

---
In reply to: [401275113](https://github.com/microsoft/onnxruntime/pull/3275#discussion_r401275113) [](ancestors = 401275113,401272474,401237498,401161112,400687293)